File: plaid-agent/src/plaid_agent/umr/penman.py
"""PENMAN notation, ported from plaid-umr's ``src/domain/format/penman.js``.

A sentence graph is written the way the UMR validator scans it, not the way a
generic PENMAN library would: a node is ``(variable / concept)`` followed by
any number of ``:relation value`` children, where a value is a child node, a
bare variable (re-entrancy), a quoted string or an atom.

Whether a bare token is a node reference or an atom cannot be decided locally,
because a reference may point forward. So the text is scanned once for
``( var /`` definitions before it is parsed, and the parser consults that set.
A token shaped like a variable that was never defined is still read as a
reference, so the caller sees the dangling edge rather than a silent atom.

This is a PORT, not an adaptation: the app writes a graph back to the same
sites it read it from, and the assistant's diff compares its own serialization
against the app's. Two readings of one text would show up as phantom changes
on a plan card.
"""

from dataclasses import dataclass, field as dc_field
from typing import Any, Dict, List, Optional, Set, Tuple

import regex as re

# Concepts, atoms and variables all stop at whitespace, brackets, a colon or
# the start of a comment.
TOKEN = re.compile(r'[^\s():#]+')

# The UMR variable convention, UFAL's own. The letter run may be non-ASCII, so
# the Unicode property escape is load-bearing.
VARIABLE = re.compile(r'^s[0-9]+\p{Ll}+[0-9]*$')

# The same, unanchored at the end: released files write ``(s6t/ thing)`` with
# no space, and the validator reads the variable off the front just like this.
VARIABLE_PREFIX = re.compile(r's[0-9]+\p{Ll}+[0-9]*')

RELATION = re.compile(r':[-A-Za-z0-9]+')
STRING = re.compile(r'"(?:\\.|[^"\\])*"')

NODE = 'node'
ATOM = 'atom'
STRING_KIND = 'string'


def is_variable(token: str) -> bool:
    """Whether a token has the shape of a UMR variable."""
    return bool(VARIABLE.match(token or ''))


@dataclass
class Child:
    rel: str
    kind: str          # 'node', 'atom' or 'string'
    value: str         # the target variable, or the literal
    inline: bool = False
    order: int = 0


@dataclass
class Node:
    var: str
    concept: str
    children: List[Child] = dc_field(default_factory=list)


@dataclass
class ParseError:
    message: str
    line: int
    col: int


@dataclass
class Graph:
    root: Optional[str] = None
    nodes: Dict[str, Node] = dc_field(default_factory=dict)
    errors: List[ParseError] = dc_field(default_factory=list)


def _mask_literals(text: str) -> str:
    """Blank out strings and comments so neither is mistaken for graph text
    while the definition set is collected. Newlines survive, so positions still
    line up."""
    out: List[str] = []
    i = 0
    n = len(text)
    while i < n:
        ch = text[i]
        if ch == '"':
            m = STRING.match(text, i)
            length = len(m.group(0)) if m else n - i
            out.append(' ' * length)
            i += length
        elif ch == '#':
            while i < n and text[i] != '\n':
                out.append(' ')
                i += 1
        else:
            out.append(ch)
            i += 1
    return ''.join(out)


def _variable_from(token: str) -> str:
    m = VARIABLE_PREFIX.match(token or '')
    return m.group(0) if m else token


def _defined_variables(masked: str) -> Set[str]:
    return {_variable_from(m.group(1)) for m in re.finditer(r'\(\s*([^\s():#]+)\s*/', masked)}


class _Scanner:
    def __init__(self, text: str):
        self.text = text
        self.i = 0
        self.line = 1
        self.col = 1

    @property
    def done(self) -> bool:
        return self.i >= len(self.text)

    def peek(self) -> str:
        return self.text[self.i] if self.i < len(self.text) else ''

    def advance(self, n: int) -> None:
        for _ in range(n):
            if self.i < len(self.text) and self.text[self.i] == '\n':
                self.line += 1
                self.col = 1
            else:
                self.col += 1
            self.i += 1

    def skip(self) -> None:
        """Whitespace and comments are the same thing to the grammar."""
        while True:
            while not self.done and self.peek().isspace():
                self.advance(1)
            if self.peek() == '#':
                while not self.done and self.peek() != '\n':
                    self.advance(1)
                continue
            return

    def here(self) -> Tuple[int, int]:
        return self.line, self.col

    def take(self, pattern) -> Optional[str]:
        m = pattern.match(self.text, self.i)
        if not m:
            return None
        self.advance(len(m.group(0)))
        return m.group(0)

    def rest(self) -> str:
        end = self.text.find('\n', self.i)
        return self.text[self.i:end if end != -1 else len(self.text)].strip()


def parse_penman(text: str) -> Graph:
    """Parse PENMAN text into a graph. Never raises: everything it cannot read
    becomes an entry in ``errors``."""
    graph = Graph()
    source = text if isinstance(text, str) else ''
    defined = _defined_variables(_mask_literals(source))
    sc = _Scanner(source)

    def fail(message: str, at: Tuple[int, int]) -> None:
        graph.errors.append(ParseError(message, at[0], at[1]))

    def read_value(children: List[Child], rel: str) -> None:
        sc.skip()
        at = sc.here()
        ch = sc.peek()
        if ch == '(':
            child = read_node()
            children.append(Child(rel, NODE, child or '', inline=True))
            return
        if ch == '"':
            raw = sc.take(STRING)
            if raw is None:
                fail(f'Unterminated string: {sc.rest()}', at)
                sc.advance(len(sc.text) - sc.i)
                return
            children.append(Child(rel, STRING_KIND, raw))
            return
        token = sc.take(TOKEN)
        if token is None:
            fail(f"Expected a value after '{rel}', found '{sc.rest() or 'end of graph'}'.", at)
            return
        if token in defined:
            children.append(Child(rel, NODE, token, inline=False))
            return
        if is_variable(token):
            fail(f"The node id (variable) '{token}' is unknown. No such node is defined.", at)
            children.append(Child(rel, NODE, token, inline=False))
            return
        children.append(Child(rel, ATOM, token))

    def read_node() -> Optional[str]:
        open_at = sc.here()
        sc.advance(1)  # the '('
        sc.skip()
        variable = sc.take(VARIABLE_PREFIX) or sc.take(TOKEN)
        if variable is None:
            fail(f"Expected a node variable id, found '{sc.rest()}'.", sc.here())
            return None
        sc.skip()
        if sc.peek() == '/':
            sc.advance(1)
            sc.skip()
        else:
            fail(f"Expected slash and concept string after '{variable}'.", sc.here())
        concept = None if sc.peek() == ')' else sc.take(TOKEN)
        if concept is None:
            fail(f"Expected a concept string for '{variable}'.", sc.here())
        node = Node(var=variable, concept=concept or '')
        if variable in graph.nodes:
            fail(f"The node id (variable) '{variable}' is not unique.", open_at)
        else:
            graph.nodes[variable] = node

        while True:
            sc.skip()
            if sc.done:
                fail(f"Graph ended without closing node '{variable}'.", sc.here())
                return variable
            ch = sc.peek()
            if ch == ')':
                sc.advance(1)
                return variable
            if ch == ':':
                at = sc.here()
                rel = sc.take(RELATION)
                if rel is None:
                    fail(f"Expected a relation label, found '{sc.rest()}'.", at)
                    sc.advance(1)
                    continue
                read_value(node.children, rel)
                continue
            # Anything else here is junk: report it once and step past the
            # whole token so the rest of the graph is still read.
            fail(f"Expected a relation or a closing bracket, found '{sc.rest()}'.", sc.here())
            if sc.take(TOKEN) is None:
                sc.advance(1)

    sc.skip()
    if sc.done:
        return graph
    if sc.peek() != '(':
        fail(f"Expected the opening bracket of the root node, found '{sc.rest()}'.", sc.here())
        return graph
    root = read_node()
    sc.skip()
    if not sc.done:
        fail(f"Unexpected content after the topmost closing bracket: '{sc.rest()}'.", sc.here())
    graph.root = root
    for node in graph.nodes.values():
        for i, child in enumerate(node.children):
            child.order = i
    return graph


def _walk(graph: Graph, visit) -> None:
    """Depth-first pre-order over child order, descending into a node the first
    time an edge reaches it. Iterative, so a pathological graph cannot blow the
    stack."""
    root, nodes = graph.root, graph.nodes
    if not root or root not in nodes:
        return
    seen = {root}
    stack = [[root, 0]]
    while stack:
        frame = stack[-1]
        node = nodes.get(frame[0])
        if node is None or frame[1] >= len(node.children):
            stack.pop()
            continue
        index = frame[1]
        frame[1] += 1
        child = node.children[index]
        if child.kind != NODE:
            continue
        already = child.value in seen
        visit(frame[0], index, child, already)
        if not already and child.value in nodes:
            seen.add(child.value)
            stack.append([child.value, 0])


def tree_edges(graph: Graph) -> Set[Tuple[str, int]]:
    """The edges at which each node is written out, by first visit."""
    first: Dict[str, Tuple[str, int]] = {}

    def visit(parent, index, child, already):
        if already or child.value in first:
            return
        first[child.value] = (parent, index)

    _walk(graph, visit)
    return set(first.values())


def _expansion_sites(graph: Graph) -> Set[Tuple[str, int]]:
    """Where each node is written out when the graph carries ``inline`` marks:
    the marked edge wins, so a graph that came from a file is written back at
    the same sites."""
    marked: Dict[str, Tuple[str, int]] = {}
    first: Dict[str, Tuple[str, int]] = {}
    any_marked = [False]

    def visit(parent, index, child, already):
        if child.inline:
            any_marked[0] = True
            marked.setdefault(child.value, (parent, index))
        if not already:
            first.setdefault(child.value, (parent, index))

    _walk(graph, visit)
    if not any_marked[0]:
        return set(first.values())
    return {marked.get(target, key) for target, key in first.items()}


def serialize_penman(graph: Optional[Graph], indent: int = 4) -> str:
    """A graph back as PENMAN, in the shape the UFAL spec shows: the root on
    the first line, every child on its own line, four spaces per level, and
    closing brackets accumulating at the end of a subtree's last line."""
    if graph is None or not graph.root or graph.root not in graph.nodes:
        return ''
    sites = _expansion_sites(graph)
    written = {graph.root}

    def pad(depth: int) -> str:
        return ' ' * (indent * depth)

    def lines(variable: str, depth: int) -> List[str]:
        node = graph.nodes[variable]
        out = [f'{pad(depth)}({variable} / {node.concept}']
        for index, child in enumerate(node.children):
            child_pad = pad(depth + 1)
            expand = (child.kind == NODE and child.value in graph.nodes
                      and child.value not in written and (variable, index) in sites)
            if expand:
                written.add(child.value)
                sub = lines(child.value, depth + 1)
                sub[0] = f'{child_pad}{child.rel} {sub[0][len(child_pad):]}'
                out.extend(sub)
            else:
                out.append(f'{child_pad}{child.rel} {child.value}')
        out[-1] += ')'
        return out

    return '\n'.join(lines(graph.root, 0))


def next_variable(sentence_index: int, concept: str, taken) -> str:
    """The next free variable for a concept in a sentence, by the standard
    rule: ``s`` + the sentence number + the concept's first letter (``x`` when
    that is not a lowercase letter) + a counter from 2 on."""
    first = (str(concept or '')[:1]).lower()
    letter = first if re.match(r'\p{Ll}', first or '') else 'x'
    base = f's{sentence_index}{letter}'
    if base not in taken:
        return base
    n = 2
    while True:
        candidate = f'{base}{n}'
        if candidate not in taken:
            return candidate
        n += 1


def graph_text(nodes: Dict[str, Node], root: Optional[str]) -> str:
    """Serialize a set of nodes from a root, marking the tree edges first (the
    same rule the canvas and the exporter use)."""
    if not root or root not in nodes:
        return ''
    graph = Graph(root=root, nodes=nodes)
    for child in (c for node in nodes.values() for c in node.children):
        child.inline = False
    for parent, index in tree_edges(graph):
        nodes[parent].children[index].inline = True
    return serialize_penman(graph)


def parse_attribute_line(line: str) -> Tuple[List[Dict[str, Any]], List[str]]:
    """``":aspect state :polarity -"`` as attribute dicts, plus what could not
    be read. The syntax is PENMAN's own, so the line is parsed as the children
    of a throwaway node rather than by a second reader of the same grammar."""
    text = (line or '').strip()
    if not text:
        return [], []
    graph = parse_penman(f'(s0x / x {text})')
    node = graph.nodes.get('s0x')
    problems = [e.message for e in graph.errors]
    attrs: List[Dict[str, Any]] = []
    for order, child in enumerate(node.children if node else []):
        if child.kind == NODE:
            problems.append(f'{child.rel} names a node, and an attribute takes a plain value.')
            continue
        attrs.append({'rel': child.rel, 'value': child.value, 'order': order})
    return attrs, problems

File: plaid-umr/src/domain/format/penman.js
// PENMAN notation: the sentence-level graph of a .umr file.
//
// The grammar is the one validate.py scans (umrtools/validate.py:623), not a
// generic PENMAN one: a node is `(variable / concept)` followed by any number
// of `:relation value` children, where a value is a child node, a bare
// variable (re-entrancy), a quoted string or an atom.
//
// Whether a bare token is a node reference or an atom cannot be decided
// locally, because a reference may point forward. So the text is scanned once
// for `( var /` definitions before it is parsed, and the parser consults that
// set. A token that is shaped like a variable but was never defined is still
// read as a reference, so the caller sees the dangling edge rather than a
// silent atom.

// Concepts, atoms and variables all stop at whitespace, brackets, a colon or
// the start of a comment (validate.py:390).
const TOKEN = /^[^\s():#]+/;

// What a concept cannot hold, then: what ends TOKEN, and a quote, which
// starts a string. Written anyway, `10:30` read back as `10` and `C#` as `C`
// with no error. The picker offers a word's form as its concept, and forms
// hold all of these.
const NOT_IN_CONCEPT = /[\s():#"]/u;

/** Why `concept` cannot be written, or null when it can. */
export const conceptProblem = (concept) =>
  NOT_IN_CONCEPT.test(concept)
    ? `A concept cannot hold spaces, brackets, colons, quotes or #: ${concept}`
    : null;

// The UMR variable convention, ÚFAL's regex (validate.py:142). The letter run
// may be non-ASCII, so the Unicode property escape is load-bearing.
const VARIABLE = /^s[0-9]+\p{Ll}+[0-9]*$/u;

// The same, unanchored at the end: released files write `(s6t/ thing)` with no
// space, and the validator reads the variable off the front just like this.
const VARIABLE_PREFIX = /^s[0-9]+\p{Ll}+[0-9]*/u;

const variableFrom = (token) => VARIABLE_PREFIX.exec(token)?.[0] ?? token;

// A relation label: a colon then letters, digits and hyphens (validate.py:391).
const RELATION = /^:[-A-Za-z0-9]+/;

const STRING = /^"(?:\\.|[^"\\])*"/;

/** Why `relation` cannot be written, or null when it can. */
export const relationProblem = (relation) => {
  const text = String(relation ?? '').trim();
  const bare = text.startsWith(':') ? text.slice(1) : text;
  if (!bare) return 'A relation needs a name after its colon.';
  return /^[-A-Za-z0-9]+$/.test(bare)
    ? null
    : `A relation holds letters, digits and hyphens only: :${bare}`;
};

/**
 * Why `value`, the value of an attribute, cannot be written, or null when it
 * can. A quoted string may hold anything; a bare atom stops where a token
 * stops, so what it cannot hold is what would be read back as something else.
 */
export const attrValueProblem = (value) => {
  const text = String(value ?? '').trim();
  if (!text) return 'An attribute needs a value.';
  if (text.startsWith('"')) {
    return STRING.test(text) && STRING.exec(text)[0] === text
      ? null
      : `A quoted value needs its closing quote: ${text}`;
  }
  if (text.includes('"')) return `A value holds a quote only around the whole of it: ${text}`;
  return NOT_IN_CONCEPT.test(text)
    ? `A value cannot hold spaces, brackets, colons or #, unless it is quoted: ${text}`
    : null;
};

// A quoted string and a comment must not be mistaken for graph text when the
// definition set is collected, so both are blanked first. Newlines survive so
// that positions still line up.
function maskLiterals(text) {
  let out = '';
  let i = 0;
  while (i < text.length) {
    const ch = text[i];
    if (ch === '"') {
      const match = STRING.exec(text.slice(i));
      const len = match ? match[0].length : text.length - i;
      out += ' '.repeat(len);
      i += len;
    } else if (ch === '#') {
      while (i < text.length && text[i] !== '\n') {
        out += ' ';
        i++;
      }
    } else {
      out += ch;
      i++;
    }
  }
  return out;
}

function definedVariables(masked) {
  const found = new Set();
  const re = /\(\s*([^\s():#]+)\s*\//g;
  let match;
  while ((match = re.exec(masked))) found.add(variableFrom(match[1]));
  return found;
}

class Scanner {
  constructor(text) {
    this.text = text;
    this.i = 0;
    this.line = 1;
    this.col = 1;
  }

  get done() {
    return this.i >= this.text.length;
  }

  peek() {
    return this.text[this.i];
  }

  advance(n) {
    for (let k = 0; k < n; k++) {
      if (this.text[this.i] === '\n') {
        this.line++;
        this.col = 1;
      } else {
        this.col++;
      }
      this.i++;
    }
  }

  // Whitespace and comments are the same thing to the grammar: a `#` runs to
  // the end of the line, and only outside a string (strings are read whole).
  skip() {
    for (;;) {
      while (!this.done && /\s/.test(this.peek())) this.advance(1);
      if (this.peek() === '#') {
        while (!this.done && this.peek() !== '\n') this.advance(1);
        continue;
      }
      return;
    }
  }

  here() {
    return { line: this.line, col: this.col };
  }

  take(re) {
    const match = re.exec(this.text.slice(this.i));
    if (!match) return null;
    const value = match[0];
    this.advance(value.length);
    return value;
  }

  /** The rest of the current line, for an error message. */
  rest() {
    const end = this.text.indexOf('\n', this.i);
    return this.text.slice(this.i, end === -1 ? this.text.length : end).trim();
  }
}

/**
 * Parse PENMAN text into a graph. Never throws: everything it cannot read
 * becomes an entry in `errors`.
 *
 * @param {string} text
 * @returns {{ root: string|null, nodes: Map<string, object>, errors: Array<{message: string, line: number, col: number}> }}
 *   Each node is `{ var, concept, children: [{ rel, kind, value, inline }] }`.
 *   `kind` is 'node' (value is the target variable, `inline` true where the
 *   child node was written out), 'string' (value keeps its quotes) or 'atom'.
 */
export function parsePenman(text) {
  const errors = [];
  const nodes = new Map();
  const source = typeof text === 'string' ? text : '';
  const defined = definedVariables(maskLiterals(source));
  const scanner = new Scanner(source);

  const fail = (message, at) => errors.push({ message, line: at.line, col: at.col });

  function readValue(children, rel) {
    scanner.skip();
    const at = scanner.here();
    const ch = scanner.peek();
    if (ch === '(') {
      const child = readNode();
      children.push({ rel, kind: 'node', value: child, inline: true });
      return;
    }
    if (ch === '"') {
      const raw = scanner.take(STRING);
      if (raw === null) {
        fail(`Unterminated string: ${scanner.rest()}`, at);
        scanner.advance(scanner.text.length - scanner.i);
        return;
      }
      children.push({ rel, kind: 'string', value: raw });
      return;
    }
    const token = scanner.take(TOKEN);
    if (token === null) {
      fail(`Expected a value after '${rel}', found '${scanner.rest() || 'end of graph'}'.`, at);
      return;
    }
    if (defined.has(token)) {
      children.push({ rel, kind: 'node', value: token, inline: false });
      return;
    }
    if (VARIABLE.test(token)) {
      fail(`The node id (variable) '${token}' is unknown. No such node is defined.`, at);
      children.push({ rel, kind: 'node', value: token, inline: false });
      return;
    }
    children.push({ rel, kind: 'atom', value: token });
  }

  function readNode() {
    const open = scanner.here();
    scanner.advance(1); // the '('
    scanner.skip();
    const variable = scanner.take(VARIABLE_PREFIX) ?? scanner.take(TOKEN);
    if (variable === null || variable === undefined) {
      fail(`Expected a node variable id, found '${scanner.rest()}'.`, scanner.here());
      return null;
    }
    scanner.skip();
    if (scanner.peek() === '/') {
      scanner.advance(1);
      scanner.skip();
    } else {
      fail(`Expected slash and concept string after '${variable}'.`, scanner.here());
    }
    const concept = scanner.peek() === ')' ? null : scanner.take(TOKEN);
    if (concept === null) {
      fail(`Expected a concept string for '${variable}'.`, scanner.here());
    }
    const node = { var: variable, concept: concept ?? '', children: [] };
    if (nodes.has(variable)) {
      fail(`The node id (variable) '${variable}' is not unique.`, open);
    } else {
      nodes.set(variable, node);
    }

    for (;;) {
      scanner.skip();
      if (scanner.done) {
        fail(`Graph ended without closing node '${variable}'.`, scanner.here());
        return variable;
      }
      const ch = scanner.peek();
      if (ch === ')') {
        scanner.advance(1);
        return variable;
      }
      if (ch === ':') {
        const at = scanner.here();
        const rel = scanner.take(RELATION);
        if (rel === null) {
          fail(`Expected a relation label, found '${scanner.rest()}'.`, at);
          scanner.advance(1);
          continue;
        }
        readValue(node.children, rel);
        continue;
      }
      // Anything else here is junk: report it once and step past the whole
      // token so the rest of the graph is still read.
      fail(`Expected a relation or a closing bracket, found '${scanner.rest()}'.`, scanner.here());
      if (scanner.take(TOKEN) === null) scanner.advance(1);
    }
  }

  scanner.skip();
  if (scanner.done) return { root: null, nodes, errors };
  if (scanner.peek() !== '(') {
    fail(
      `Expected the opening bracket of the root node, found '${scanner.rest()}'.`,
      scanner.here(),
    );
    return { root: null, nodes, errors };
  }
  const root = readNode();
  scanner.skip();
  if (!scanner.done) {
    fail(
      `Unexpected content after the topmost closing bracket: '${scanner.rest()}'.`,
      scanner.here(),
    );
  }
  return { root: root ?? null, nodes, errors };
}

const edgeKey = (parent, index) => `${parent}\u0000${index}`;

// Depth-first pre-order over child order, descending into a node the first
// time an edge reaches it. Iterative rather than recursive only so that a
// pathological graph cannot blow the stack.
function walk(graph, visit) {
  const { root, nodes } = graph;
  if (!root || !nodes.has(root)) return;
  const seen = new Set([root]);
  const stack = [{ variable: root, next: 0 }];
  while (stack.length) {
    const frame = stack[stack.length - 1];
    const node = nodes.get(frame.variable);
    if (!node || frame.next >= node.children.length) {
      stack.pop();
      continue;
    }
    const index = frame.next++;
    const child = node.children[index];
    if (child.kind !== 'node') continue;
    const alreadySeen = seen.has(child.value);
    visit(frame.variable, index, child, alreadySeen);
    if (!alreadySeen && nodes.has(child.value)) {
      seen.add(child.value);
      stack.push({ variable: child.value, next: 0 });
    }
  }
}

/**
 * The edges at which each node is written out, by first visit: a depth-first
 * walk from the root following child order, expanding a node the first time an
 * edge reaches it and writing a bare variable everywhere else.
 *
 * @param {{root: string|null, nodes: Map}} graph
 * @returns {Set<[string, number]>} (parent variable, child index) pairs
 */
export function treeEdges(graph) {
  const first = new Map();
  walk(graph, (parent, index, child, alreadySeen) => {
    if (alreadySeen || first.has(child.value)) return;
    first.set(child.value, [parent, index]);
  });
  return new Set(first.values());
}

// Where each node is written out when the graph carries `inline` markers: the
// marked edge wins, so a graph that came from a file is written back at the
// same sites. A node whose marked edge is unreachable still gets written at
// the first edge that reaches it.
function expansionSites(graph) {
  const marked = new Map();
  const first = new Map();
  let anyMarked = false;
  walk(graph, (parent, index, child, alreadySeen) => {
    if (child.inline === true) {
      anyMarked = true;
      if (!marked.has(child.value)) marked.set(child.value, edgeKey(parent, index));
    }
    if (!alreadySeen && !first.has(child.value)) first.set(child.value, edgeKey(parent, index));
  });
  if (!anyMarked) return new Set(first.values());
  const sites = new Set();
  for (const [target, key] of first) sites.add(marked.get(target) ?? key);
  return sites;
}

/**
 * Write a graph back as PENMAN, in the canonical shape the ÚFAL spec shows:
 * the root on the first line, every child on its own line, four spaces of
 * indentation per level, and closing brackets accumulating at the end of the
 * last line of a subtree.
 *
 * @param {{root: string|null, nodes: Map}} graph
 * @param {{indent?: number}} [options]
 * @returns {string}
 */
export function serializePenman(graph, options = {}) {
  const indent = options.indent ?? 4;
  const { root, nodes } = graph || {};
  if (!root || !nodes || !nodes.has(root)) return '';
  const sites = expansionSites(graph);
  const written = new Set([root]);

  const pad = (depth) => ' '.repeat(indent * depth);

  function lines(variable, depth) {
    const node = nodes.get(variable);
    const out = [`${pad(depth)}(${variable} / ${node.concept}`];
    node.children.forEach((child, index) => {
      const childPad = pad(depth + 1);
      const expandHere =
        child.kind === 'node' &&
        nodes.has(child.value) &&
        !written.has(child.value) &&
        sites.has(edgeKey(variable, index));
      if (expandHere) {
        written.add(child.value);
        const sub = lines(child.value, depth + 1);
        sub[0] = `${childPad}${child.rel} ${sub[0].slice(childPad.length)}`;
        out.push(...sub);
      } else {
        out.push(`${childPad}${child.rel} ${child.value}`);
      }
    });
    out[out.length - 1] += ')';
    return out;
  }

  return lines(root, 0).join('\n');
}
